### ISSUE TITLE
Error concatenating grouped bar charts: Javascript Error: Undefined data set name 

### ISSUE THREAD

**Issue by zhanjiezhu** [altair-viz/altair#1727]
When I tried to concatenate two grouped bar charts, it throws error "Javascript Error: Undefined data set name: "scale_concat_1_child_main"

Example can be reproduced using the grouped bar chart example (https://altair-viz.github.io/gallery/grouped_bar_chart.html)
```python
import altair as alt
from vega_datasets import data
source = data.barley()
test = alt.Chart(source).mark_bar().encode(
    x='year:O',
    y='sum(yield):Q',
    color='year:N',
    column='site:N'
)
test & test
```

**Comment by jakevdp** [comment]
This is a known bug in Vega-Lite; see https://github.com/vega/vega-lite/issues/5261

**Comment by torresweb** [comment]
Hi, I am using vega-lite 4.13.1 and with the same problem. I am not sure if #2760 should fix it or not. Any workaround?

**Comment by jakevdp** [comment]
vega/vega-lite#2760 has not been fixed, so it is still an issue.

**Comment by jesnie** [comment]
I'm running into a problem that looks very much like this. I must admit I don't understand why #2760 relates to this. #2760 talks about generic specs inside facets, but I want to concatenate bar charts. I don't suppose you can give me a hint about what's going on here, so that I might work around it?

**Comment by jakevdp** [comment]
I don't think there's any workaround: Vega-Lite currently does not support concatenation of some compound charts.

**Comment by torresweb** [comment]
I found a workaround for that. You need to create each group of bar columns with a graph and then concatenate these graphs, sharing Y scale and hiding/unhiding this axis.

**Comment by ibayer-ny** [comment]
@torresweb 

I don't understand your workaround description. Could you please provide an example?
Maybe transforming the initial example in this issue would be a quick way to demonstrate it.

**Comment by joelostblom** [comment]
Details of why this is not working can be found in https://github.com/vega/vega-lite/issues/4680, specifically [this comment](https://github.com/vega/vega-lite/issues/4680#issuecomment-782952466). It only happens when VegaLite tries to reference the same data set in two concatenated charts. A workaround is to resort the data (e.g. `df.sample(df.shape[0])`), so that VegaLite creates to separate data source references:

```py
import altair as alt
from vega_datasets import data
source = data.barley()
test = alt.Chart(source).mark_bar().encode(
    x='year:O',
    y='sum(yield):Q',
    color='year:N',
    column='site:N'
)
test & test.properties(data=source.sample(source.shape[0]))
```

![image](https://user-images.githubusercontent.com/4560057/108643193-4b0fb300-745e-11eb-9ad5-c2434e678d3e.png)

The obvious caveat is that if you are doing something that depends on the order of your data, this is not a good idea, but for concatenated charts this should be less of an issue as long as the dimensions are shared.

@zhanjiezhu Would you mind closing this issues as it is already tracked upstreams in Vega-Lite in the linked issues and there is no fix needed in Altair?

**Comment by joelostblom** [comment]
This will be fixed once https://github.com/altair-viz/altair/pull/2871 is merged

**Comment by binste** [comment]
Fixed in Altair 5.
